Automattic/wp-calypso
# Add user store to packages/data-stores and use wpcom-proxy-request for API calls

The object of this PR is to add a `user` data store that handles checking for a logged in user, and account creation with fetching and error states, using patterns from the `@wordpress/data` package. A number of selectors and actions are provided to hopefully allow for creating simple sign up interfaces. I've included a readme with an example of how to use the data store.

#### Changes proposed in this Pull Request

* Add `wpcom-proxy-request` to `packages/data-stores` for making proxied API calls to wp.com
* Add a simple `currentUser` object
* Add a simple `newUser` object
* Add a Promise-based wrapper for `wpcom-proxy-request` calls [similar to what's used in `wpcom.js`](https://github.com/Automattic/wpcom.js/blob/94edca17553d597f9de68fc03efe488158a5babd/lib/util/send-request.js#L88)
* Add `debug`
* Add actions and selectors for checking if a user is logged in, and for creating a new user using passwordless login, with error and fetching state

This PR supersedes the other similar experiments in #38679 and #38739

#### Still to be done

* [x] ~Add fetching state~
* [x] ~Clean up types a bit more (error object is currently set to `object`)~
* [x] ~Remove magic strings (put the client id and 'secret' somewhere else)~

#### Testing instructions



* This PR adds an extra data store but doesn't yet use it anywhere so there are no regressions to test against.
* To test the behaviour of the store, follow the included `README.md` file which has an example signup form you can use.
* To see debug logs of the `wpcom-proxy-request` responses, run the following in your browser console `localStorage.setItem( 'debug', 'data-stores:*' );`

Fixes #38794
